### PR TITLE
Add `CallbackHandler`, `CallbackManager` and `Tracer` with tests

### PR DESCRIPTION
**Pull request by agola11**
Add callback handlers, manager and tracer implementation with tests.

Has not been wired up to base classes yet! Will do that in a followup PR.

### CODE REVIEW

File: langchain/src/callbacks/index.ts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    tool: { name: string },
+    input: string,
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleToolError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleToolEnd?(output: string, verbose?: boolean): Promise<void>;
+
+  handleText?(text: string, verbose?: boolean): Promise<void>;
```
Comment (nfcampos) [inline langchain/src/callbacks/index.ts:65]: Remind me when this one is fired?

File: langchain/src/callbacks/index.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  async handleLLMStart(
+    llm: { name: string },
+    prompts: string[],
+    verbose?: boolean
+  ): Promise<void> {
+    for (const handler of this.handlers) {
+      if (!handler.ignoreLLM && (verbose || handler.alwaysVerbose)) {
+        await handler.handleLLMStart?.(llm, prompts);
```
Comment (nfcampos) [inline langchain/src/callbacks/index.ts:99]: How about adding a try catch around each call to a handler. That way of one handler errors your chain run (or whatever it was) doesn't fail, and a failure in one handler doesn't prevent calling the next handlers in the list

File: langchain/src/callbacks/index.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+  addHandler(handler: BaseCallbackHandler): void {
+    this.handlers.push(handler);
+  }
+
+  removeHandler(handler: BaseCallbackHandler): void {
+    this.handlers = this.handlers.filter((_handler) => _handler !== handler);
+  }
+
+  setHandlers(handlers: BaseCallbackHandler[]): void {
```
Comment (nfcampos) [inline langchain/src/callbacks/index.ts:219]: Maybe this could be named replaceHandlers to be clearer?

File: langchain/src/callbacks/index.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    this.handlers = handlers;
+  }
+}
+
+export class ConsoleCallbackHandler extends BaseCallbackHandler {
+  constructor() {
```
Comment (nfcampos) [inline langchain/src/callbacks/index.ts:225]: This isn't needed (when it's empty)

**Review by nfcampos — COMMENTED** [review]
Some comments, looks good

File: langchain/src/callbacks/index.ts (hunk 1 above)
Comment (agola11) [reply langchain/src/callbacks/index.ts:65]: This is used whenever we need to log arbitrary messages from chains or agents that are not necessarily on start/end events. TBH I'm not in love with this, and want to look into another soln but it's not a priority rn, so I just copied what's in python

File: langchain/src/callbacks/index.ts (hunk 2 above)
Comment (agola11) [reply langchain/src/callbacks/index.ts:99]: Great idea

Comment (agola11) [reply langchain/src/callbacks/index.ts:99]: resolved in followup PR

**Pull request merged**